temporal-test-server/src/test/java/io/temporal/testserver/functional/activity/ActivityHeartbeat.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.testserver.functional.activity;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+
+import com.google.protobuf.ByteString;
+import io.temporal.activity.Activity;
+import io.temporal.activity.ActivityInfo;
+import io.temporal.activity.ActivityOptions;
+import io.temporal.api.common.v1.Payloads;
+import io.temporal.api.workflowservice.v1.RecordActivityTaskHeartbeatRequest;
+import io.temporal.common.RetryOptions;
+import io.temporal.common.converter.DefaultDataConverter;
+import io.temporal.failure.ActivityFailure;
+import io.temporal.testing.internal.SDKTestWorkflowRule;
+import io.temporal.testserver.functional.common.TestActivities;
+import io.temporal.testserver.functional.common.TestWorkflows;
+import io.temporal.workflow.Workflow;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class ActivityHeartbeat {
+  private static final ConcurrentLinkedQueue<Optional<Payloads>> activityHeartbeats =
+      new ConcurrentLinkedQueue<>();
+
+  @Rule
+  public SDKTestWorkflowRule testWorkflowRule =
+      SDKTestWorkflowRule.newBuilder()
+          .setWorkflowTypes(TestWorkflow.class)
+          .setActivityImplementations(new TestActivity())
+          .build();
+
+  @Test
+  public void testActivityHeartbeatNoLastHeartbeatDetails() {
+    // Test that when last heartbeat details are not set on failure, the test server
+    // clear the heartbeat details.
+    String result =
+        testWorkflowRule.newWorkflowStub(TestWorkflows.WorkflowReturnsString.class).execute();
+    assertEquals("", result);
+    assertEquals(2, activityHeartbeats.size());
+    assertFalse(activityHeartbeats.poll().isPresent());
+    assertEquals(
+        "heartbeat details",
+        DefaultDataConverter.STANDARD_INSTANCE.fromPayloads(
+            0, activityHeartbeats.poll(), String.class, String.class));
```

Comment:
I assume this is the assertion that would have failed without the change because it'd still have 2 in the queue, but even the second would not be present due to overwrite (just kinda proving I read it and understand it heh)

Reply:
Yes this fails without the test server change